services/web/server/tests/unit/with_dbs/04/functions_rpc/test_functions_controller_rpc.py

```diff
@@ -0,0 +1,622 @@
+    assert saved_function.project_id == mock_function.project_id
+
+    # Assert the returned function matches the expected result
+    assert registered_function.title == mock_function.title
+    assert registered_function.description == mock_function.description
+    assert isinstance(registered_function, ProjectFunction)
+    assert registered_function.project_id == mock_function.project_id
+
+
+async def test_get_function(
+    client: TestClient, rpc_client: RabbitMQRPCClient, mock_function: ProjectFunction
+):
+    assert client.app
+    # Register the function first
+    registered_function = await functions_rpc.register_function(
+        rabbitmq_rpc_client=rpc_client, function=mock_function
+    )
+    assert registered_function.uid is not None
+
+    # Retrieve the function using its ID
+    retrieved_function = await functions_rpc.get_function(
+        rabbitmq_rpc_client=rpc_client, function_id=registered_function.uid
+    )
+
+    # Assert the retrieved function matches the registered function
+    assert retrieved_function.uid == registered_function.uid
+    assert retrieved_function.title == registered_function.title
+    assert retrieved_function.description == registered_function.description
+
+    # Ensure retrieved_function is of type ProjectFunction before accessing project_id
+    assert isinstance(retrieved_function, ProjectFunction)
+    assert isinstance(registered_function, ProjectFunction)
+    assert retrieved_function.project_id == registered_function.project_id
+
+
+async def test_get_function_not_found(
+    client: TestClient, rpc_client: RabbitMQRPCClient
+):
+    assert client.app
+    # Attempt to retrieve a function that does not exist
+    with pytest.raises(FunctionIDNotFoundError):
+        await functions_rpc.get_function(
+            rabbitmq_rpc_client=rpc_client, function_id=uuid4()
+        )
+
+
+async def test_list_functions(client: TestClient, rpc_client: RabbitMQRPCClient):
+    assert client.app
+    # Register a function first
+    mock_function = ProjectFunction(
+        title="Test Function",
+        description="A test function",
+        input_schema=JSONFunctionInputSchema(),
+        output_schema=JSONFunctionOutputSchema(),
+        project_id=uuid4(),
+        default_inputs=None,
+    )
+    registered_function = await functions_rpc.register_function(
+        rabbitmq_rpc_client=rpc_client, function=mock_function
+    )
+    assert registered_function.uid is not None
+
+    # List functions
+    functions, _ = await functions_rpc.list_functions(
+        rabbitmq_rpc_client=rpc_client, pagination_limit=10, pagination_offset=0
+    )
+
+    # Assert the list contains the registered function
+    assert len(functions) > 0
+    assert any(f.uid == registered_function.uid for f in functions)
+
+
+async def test_list_functions_empty(
+    client: TestClient, rpc_client: RabbitMQRPCClient, clean_functions: None
+):
+    assert clean_functions is None
```

Comment:
you don't need this, if the tooling complains use # pylint: disable=....

Reply:
done